homeassistant/components/template/binary_sensor.py

```diff
@@ -301,9 +307,36 @@ def __init__(
                 self._parse_result.add(key)
 
         self._delay_cancel: CALLBACK_TYPE | None = None
-        self._auto_off_cancel = None
+        self._auto_off_cancel: CALLBACK_TYPE | None = None
+        self._auto_off_time: datetime | None = None
         self._state: bool | None = None
 
+    async def async_added_to_hass(self) -> None:
+        """Restore last state."""
+        await super().async_added_to_hass()
+        if (
+            (last_state := await self.async_get_last_state()) is not None
+            and (extra_data := await self.async_get_last_binary_sensor_data())
+            is not None
+            and last_state.state not in (STATE_UNKNOWN, STATE_UNAVAILABLE)
+            # The trigger might have fired already while we waited for stored data,
+            # then we should not restore state
+            and self._state is None
+        ):
+            self._state = last_state.state == STATE_ON
+
+            if CONF_AUTO_OFF not in self._config:
+                return
+
+            if (
+                auto_off_time := extra_data.auto_off_time
+            ) is not None and auto_off_time <= dt_util.utcnow():
+                # It's already past the saved auto off time
+                self._state = False
+
+            if self._state and auto_off_time is not None:
+                self._set_auto_off(auto_off_time)
```

Comment:
```suggestion
            if self._state and auto_off_time is not None and self._auto_off_time is None:
                self._set_auto_off(auto_off_time)
```
The same race condition you identified above also occurs here. If the trigger fired while you were waiting for stored data then you don't want to restore its timer here.